velox/functions/prestosql/aggregates/ChecksumAggregate.cpp

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#define XXH_INLINE_ALL
+
+#include "velox/exec/Aggregate.h"
+#include "velox/expression/FunctionSignature.h"
+#include "velox/external/xxhash.h"
+#include "velox/functions/prestosql/aggregates/AggregateNames.h"
+#include "velox/functions/prestosql/aggregates/PrestoHasher.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::aggregate {
+
+namespace {
+
+/// Computes an order-insensitive checksum of the input
+/// vector.
+///
+/// checksum(T)-> varbinary
+class ChecksumAggregate : public exec::Aggregate {
+ public:
+  explicit ChecksumAggregate(const TypePtr& resultType)
+      : Aggregate(resultType) {}
+
+  void finalize(char** /* groups */, int32_t /* numGroups */) override {}
+
+  int32_t accumulatorFixedWidthSize() const override {
+    return sizeof(int64_t);
+  }
+
+  void initializeNewGroups(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    setAllNulls(groups, indices);
+    for (auto i : indices) {
+      *value<int64_t>(groups[i]) = 0;
+    }
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    auto* vector = (*result)->asUnchecked<FlatVector<StringView>>();
+    VELOX_CHECK(vector);
+    vector->resize(numGroups);
+
+    auto* rawValues = vector->mutableRawValues();
+    for (auto i = 0; i < numGroups; ++i) {
+      auto group = groups[i];
+      if (isNull(group)) {
+        vector->setNull(i, true);
+      } else {
+        vector->setNull(i, false);
```

Comment:
In most cases, there will be no nulls. Hence, it would be more efficient to clear all nulls before the loop and call setNull only if isNull(group).

```
vector->clearAllNulls();
for (...) {

   if (isNull(group)) {
      vector->setNull(i, true);
   } else {
       rawValues[i] =...
   }
}
```